Tower workflow schema

sean-m-sullivan opened the pull request:
##### SUMMARY
Adds a way to add entire workflow node schemas to workflows. Either through the workflow schema module or the workflow job template module. 

This speeds up workflow creation vs the workflow node module by 3x. 
The model for the schemas is the format used by the tower_export module. 

The main difference between this and the workflow node module is that the loops are done in python. Traditionally if you have a workflow with 10 nodes, ansible tasks need to be invoked 19 times. 1x to create the workflow, 10 x to initially create the nodes, and then one time for each node that is not an endpoint in the schema. This removes the need to loop and invoke many times. 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
awx-collection

##### AWX VERSION
17.0.1

sean-m-sullivan committed:
add tower workflow schema

sean-m-sullivan committed:
add tower workflow schema

xstasi committed:
Add tower group for collection modules

sean-m-sullivan committed:
Made changes to work with inventory sources

Made changes to work with inventory sources

sean-m-sullivan committed:
Made changes to work with inventory sources

Made changes to work with inventory sources

sean-m-sullivan committed:
update documentation

sean-m-sullivan committed:
update documentation

sean-m-sullivan committed:
update documentation

sean-m-sullivan committed:
update documentation

sean-m-sullivan committed:
added tests, additionial documentation, fixed search bug

sean-m-sullivan committed:
added tests, additionial documentation, fixed search bug

sean-m-sullivan committed:
updated documentation fixed lint

sean-m-sullivan committed:
updated documentation fixed lint

sean-m-sullivan committed:
Merge pull request #9 from ansible/devel

Rebase from devel

sean-m-sullivan committed:
update to recent changes

sean-m-sullivan committed:
update to recent changes

sean-m-sullivan committed:
Merge pull request #11 from ansible/devel

Rebase

sean-m-sullivan committed:
api rebase

sean-m-sullivan committed:
Merge pull request #25 from ansible/devel

Rebase

sean-m-sullivan committed:
update to use new exit

sean-m-sullivan committed:
update to use new exit

sean-m-sullivan committed:
Merge pull request #40 from ansible/devel

Rebase

sean-m-sullivan committed:
update to add schemas to workflow template

sean-m-sullivan committed:
update to add schemas to workflow template

sean-m-sullivan committed:
update docs

sean-m-sullivan committed:
update docs

sean-m-sullivan committed:
fix typo

sean-m-sullivan committed:
fix typo

sean-m-sullivan committed:
fix completeness parameters

sean-m-sullivan committed:
fix completeness parameters

sean-m-sullivan committed:
linting

sean-m-sullivan committed:
linting

sean-m-sullivan committed:
linting

sean-m-sullivan committed:
linting

beeankha commented on awx_collection/plugins/modules/tower_workflow_job_template_schema.py:
```diff
@@ -0,0 +1,347 @@
+# coding: utf-8 -*-
+
+
+# (c) 2020, John Westcott IV <john.westcott.iv@redhat.com>, Sean Sullivan <ssulliva@redhat.com>
```
Change :copyright: year to 2021

sean-m-sullivan replied:
Updated.

sean-m-sullivan committed:
update year

sean-m-sullivan committed:
update year

sean-m-sullivan committed:
fix linting

sean-m-sullivan committed:
fix linting

sean-m-sullivan committed:
fix logic

sean-m-sullivan committed:
fix logic

kdelee commented on awx_collection/plugins/modules/tower_workflow_job_template.py:
```diff
@@ -142,11 +300,66 @@
+          name: example-project
+          inventory:
+            organization:
+              name: Default
```
is the indentation correct here? looks a bit wonky
I don't understand why the inventory looks like a sub-item of unified_job_template or why organization is a sub item of inventory

kdelee replied:
:mag: ok I think I get it....you are trying to indicate the organization of the unified job template indirectly by saying use the JT with name `example-project` that has an inventory (not named) that is in the Default organization.

Is that right?

sean-m-sullivan replied:
This is for an inventory sync update in a workflow. It's actually the direct output of tower_export and the way it keeps it in the api. The goal is to reduce looping to create nodes of a workflow, and it follows the model from tower_export, which follows how it keeps things in the API. The fact that organization is referenced differently for inventories vs Job templates is just the way current tower api works.

sean-m-sullivan committed:
update

sean-m-sullivan committed:
update

john-westcott-redhat commented:
@sean-m-sullivan, How does tower_workflow_job_template_schema.py compare to the existing tower_workflow_job_template_node.py? Also, if there is a module for managing the schema why support it in tower_workflow_job_template?

sean-m-sullivan commented:
> How does tower_workflow_job_template_schema.py compare to the existing tower_workflow_job_template_node.py? Also, if there is a module for managing the schema why support it in tower_workflow_job_template?

TBH I added the separate module just to add options, it could be done with just the existing tower_workflow_job_template. 

In terms of comparison to tower_workflow_job_template_node, the difference is that node does a single node creation, link, OR modify existing settings/links, The code is much the same, except it adds the ability to loop over nodes, create or make sure they exist, then link them, in a single ansible module for an entire schema. As mentioned above, its meant to reduce 18 calls to tower_workflow_job_template_node and a call to tower_workflow_job_template to a single call,

sean-m-sullivan committed:
Merge pull request #49 from ansible/devel

Rebase

keithjgrant committed:
add VariablesField YAML/JSON validation

marshmalien committed:
Add subscription wizard and redirect logic

sean-m-sullivan committed:
Merge branch 'devel_copy_workflow_schema' into tower_workflow_schema

sean-m-sullivan committed:
update logic

sean-m-sullivan committed:
update logic

sean-m-sullivan committed:
Merge branch 'tower_workflow_schema' of github.com:sean-m-sullivan/awx into tower_workflow_schema

ryanpetrello commented:
@beeankha @john-westcott-iv is this PR ready for another set of eyeballs, or is there still unaddressed feedback?

sean-m-sullivan commented:
@ryanpetrello I had talked with Beeankha about this, and we were waiting on Alan Coding to come back to do a review, in addition I found an bug where we didn't address approval nodes, and will try to get that in. In addition, the copy PR #9374 somehow got mixed in here, and I need to extract it.

sean-m-sullivan committed:
Merge pull request #53 from ansible/devel

Rebase

sean-m-sullivan committed:
fix approval node logic

sean-m-sullivan committed:
fix approval node logic

sean-m-sullivan committed:
:Merge branch 'tower_workflow_schema' of github.com:sean-m-sullivan/awx into tower_workflow_schema

sean-m-sullivan committed:
Merge pull request #54 from ansible/devel

Rebase

sean-m-sullivan commented:
Regate

sean-m-sullivan committed:
update documentation

sean-m-sullivan committed:
update documentation

sean-m-sullivan committed:
black linting

jborean93 committed:
Fix up request tower ps script

jborean93 committed:
Missed ErrorAction in the params

nixocio committed:
Remove isolated and controller from the UI

Remove isolated and controller from the UI.

See: https://github.com/ansible/awx/issues/9752

AlexSCorey committed:
Updates Lingui

sean-m-sullivan committed:
update to fix approval bug

sean-m-sullivan committed:
update to fix approval bug

sean-m-sullivan committed:
Merge pull request #63 from ansible/devel

Rebase

beeankha commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -38,7 +38,7 @@
     'tower_job_cancel',
     'tower_workflow_template',
-    'tower_ad_hoc_command_wait',
+    'tower_workflow_job_template_schema' 'tower_ad_hoc_command_wait',
```
Looks like there's a missing `,` and newline after `'tower_workflow_job_template_schema'`

beeankha commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -59,8 +59,8 @@
-    'tower_workflow_job_template': ['survey_spec'],
-    # ad hoc commands support interval and timeout since its more like tower_job_launch
+    'tower_workflow_job_template': ['survey_spec', 'destroy_current_schema'],
+    # ad hoc commands support interval and timeout since its more like tower_job_launc
```
An `h` got deleted at the end of this comment line

sean-m-sullivan committed:
update test completeness

sean-m-sullivan committed:
update test completeness

nixocio committed:
Do not validate optional survey

Do not validate optional survey.

See: https://github.com/ansible/awx/issues/9176

sean-m-sullivan committed:
update approval create logic

sean-m-sullivan committed:
update approval create logic

sean-m-sullivan committed:
Merge branch 'tower_workflow_schema' of github.com:sean-m-sullivan/awx into tower_workflow_schema

shanemcd committed:
Cluster dev env: only run migrations on first node

jbradberry committed:
Adjust datetimes to be aware when using awx-manage gather_analytics

Also, make sure that an explicit `since` parameter will win over
default `until=now()` when calculating the 4-week data limit.

AlexSCorey committed:
updates the method of loading locales

sean-m-sullivan committed:
Merge pull request #65 from ansible/devel

Rebase

jakemcdermott committed:
Limit the visible playbook select region

softwarefactory-project-zuul[bot] committed:
Merge pull request #9827 from ansible/jakemcdermott-playbook-max-size

Limit the visible playbook select region

SUMMARY
before

after

Reviewed-by: Kersom <None>

AlexSCorey committed:
updates importing messages properly

softwarefactory-project-zuul[bot] committed:
Merge pull request #9673 from jborean93/pwsh-request

Fix up request tower ps script

SUMMARY
Fixes the request_tower_configuration.ps1 script with the following:

Added PowerShell 6+ support, can now run on Linux
Removes the uneeded strict and ErrorActionPreference settings which could impact the global settings
Enables TLS 1.2 support on Windows PowerShell
Fixes -insecure handling for certificates

Works across ps versions
Original didn't actually work (at least in my testing)
Hopefully doesn't impact the global state


Fixes IE not loaded error on Windows PowerShell when running on a new profile
Changes the encoding to UTF-8 (no BOM) and made it executable

Can now be invoked on Linux like ./request_tower_configuration.ps1 ...
There are no non-ASCII chars in the script so this won't matter on Windows



Fixes #9434
ISSUE TYPE

Bugfix Pull Request

COMPONENT NAME

Tools

AWX VERSION
N/A

Reviewed-by: Ryan Petrello <None>

ryanpetrello committed:
automatically setup default Galaxy credentials on SAML login

sean-m-sullivan committed:
update schema unified job template types in tests

sean-m-sullivan committed:
update schema unified job template types in tests

sean-m-sullivan committed:
Merge branch 'tower_workflow_schema' of github.com:sean-m-sullivan/awx into tower_workflow_schema

softwarefactory-project-zuul[bot] committed:
Merge pull request #9825 from jbradberry/analytics-naive-datetimes-fix

Adjust datetimes to be aware when using awx-manage gather_analytics

SUMMARY
Adjust datetimes to be aware when using awx-manage gather_analytics
Also, make sure that an explicit since parameter will win over
default until=now() when calculating the 4-week data limit.
ISSUE TYPE

Bugfix Pull Request

COMPONENT NAME

API

AWX VERSION

Reviewed-by: Ryan Petrello <None>

jakemcdermott committed:
Use shared logo img and dynamic copyright date

sean-m-sullivan committed:
update schema unified job template types in tests

sean-m-sullivan committed:
update schema unified job template types in tests

jakemcdermott committed:
Ignore no-literal rule for AboutModal component

shanemcd committed:
Update version and changelog for version 19

softwarefactory-project-zuul[bot] committed:
Merge pull request #9803 from shanemcd/cluster-migrate-once

Cluster dev env: only run migrations on first node

Reviewed-by: Ryan Petrello <None>
Reviewed-by: Shane McDonald <me@shanemcd.com>

tiagodread committed:
Remove duple //

jakemcdermott committed:
Update CHANGELOG.md

softwarefactory-project-zuul[bot] committed:
Merge pull request #9496 from marshmalien/settings-subscription-wizard

Subscription wizard

SUMMARY

Adds subscriptions routes
Extends Config context to return an array with two items: config state and config setter

Update components that use Config with the new return pattern


Move mock config into setupTests.js
Return only the routes the user is "authorized" (valid license key) to view

Subscription Details view: /settings/subscription/details

This is our standard details view
Clicking Edit will send the user to subscription wizard view /settings/subscription/edit
Route is not accessible when license type is OPEN

Subscription Add wizard view: /settings/subscription_management
Step 1 - Subscription:

If a user does not have a Red Hat Ansible Automation Platform subscription, they can request a trial subscription via the link
Toggle between uploading a subscription manifest .zip file or retrieving subscriptions using Red Hat credentials (username and password)
Get Subscriptions button fetches subscriptions and displays them in a modal



Step 2 - Tracking and analytics:

Shows two checkboxes to enable User analytics and Automation analytics
If the user has previously selected the RH subscription manifest flow, checking the Automation Analytics box will display required RH username and password fields
If the user has previously selected the RH username/password flow, they will not see this additional username/password field if Automation Analytics is checked


Step 3 - EULA: https://tower-mockups.testing.ansible.com/patternfly/settings/settings-license-step-03/

Submission should show a success message and navigate user to dashboard if this is the initial launch and to the subscription detail view if they are editing the subscription
Failed submission should show a wizard form error

ISSUE TYPE

Feature

COMPONENT NAME

UI

ADDITIONAL INFORMATION

Reviewed-by: Michael Abashian <None>
Reviewed-by: Kersom <None>
Reviewed-by: Tiago Góes <tiago.goes2009@gmail.com>

nixocio committed:
Fix issue related to plural

Fix issue related to plural

softwarefactory-project-zuul[bot] committed:
Merge pull request #9843 from tiagodread/fix-doc-link

Remove duple // for docs link on PageHeaderToolBar

Related: #9842

Reviewed-by: Jake McDermott <yo@jakemcdermott.me>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9836 from jakemcdermott/fix-9830-9296

Use shared logo img and dynamic copyright date

for #9830, #9296

Reviewed-by: Kersom <None>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9844 from nixocio/ui_fix_plural_issue

Fix issue related to plural

Fix issue related to plural
Manually tested. make ui-devel

Reviewed-by: Jake McDermott <yo@jakemcdermott.me>

fosterseth committed:
Fix elapsed time on job showing incorrect value

- job elapsed time showed 0.0, during and after the job run

softwarefactory-project-zuul[bot] committed:
Merge pull request #9840 from shanemcd/bump-19

Update version and changelog for version 19

Reviewed-by: Jake McDermott <yo@jakemcdermott.me>
Reviewed-by: Ryan Petrello <None>
Reviewed-by: Bianca Henderson <beeankha@gmail.com>

marshmalien committed:
Update subscription docs link to use single source of truth

fosterseth committed:
string

softwarefactory-project-zuul[bot] committed:
Merge pull request #9848 from marshmalien/subscription-docs-link

Update subscription docs link to use single source of truth

SUMMARY
Related: #8428

Update subscription edit form documentation links to use the docs link utility.
Fix broken unit tests

COMPONENT NAME


UI

Reviewed-by: Keith Grant <None>

fosterseth committed:
the returned self.elapsed should be decimal, instead of float, to match django model field type

fosterseth committed:
fix awx_collections tests by comparing Decimal to float, instead of comparing strings

softwarefactory-project-zuul[bot] committed:
Merge pull request #9755 from nixocio/ui_issue_9752

Remove isolated and controller from the UI

Remove isolated and controller from the UI.
See: #9752

Reviewed-by: Jake McDermott <yo@jakemcdermott.me>

marshmalien committed:
Add remaining expanded template list item details

jakemcdermott committed:
Fix non-translatable template string

john-westcott-iv commented:
With the changes you've made on this do we still need to introduce the new module tower_workflow_job_template_schema.py or can we just have tower_workflow_job_template? If we still need the second one, can you summarize the difference between them?

sean-m-sullivan commented:
@john-westcott-iv 
Part of it was that I created the tower_workflow_job_template_schema for testing, and its meant to mimic how people use the tower_workflow_job_template_node, where it only edits the schema. I am in favor of just removing it. I kept it, on the off chance it was decided to only have it as its own module.

We can just have tower_workflow_job_template changes. The goal was to make it functionally work like tower_workflow_template that we are depreciating, and to put the looping needed to create a schema in python rather then what is the current method of using ansible loops.

sean-m-sullivan committed:
update to remove seperate schema module

sean-m-sullivan committed:
update to remove seperate schema module

jakemcdermott committed:
Use validated prop in cred plugin field

softwarefactory-project-zuul[bot] committed:
Merge pull request #9869 from jakemcdermott/fix-cred-validation-prop

Use validated prop in cred plugin field

SUMMARY
Fxes an initialization bug for cred plugin fields, introduced by this commit
This pr applies the patch described below:
cc @marshmalien @unlikelyzero

before


after

Reviewed-by: John Hill <johill@redhat.com>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9861 from jakemcdermott/fix-9296

Fix non-translatable template string

SUMMARY
The template string for the trademark doesn't seem like it can be used for a valid translation key. This caused the raw variable to be displayed for production builds instead of a readable string.  see: #9296 (comment)
ADDITIONAL INFORMATION
Screenshots, after applying this patch:
development environment


production build

Reviewed-by: Kersom <None>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9814 from nixocio/ui_issue_9176

Do not validate optional survey

Do not validate optional survey.
See: #9176

Reviewed-by: Marliana Lara <marliana.lara@gmail.com>
Reviewed-by: Tiago Góes <tiago.goes2009@gmail.com>

john-westcott-iv commented on awx_collection/test/awx/test_completeness.py:
```diff
@@ -38,6 +38,7 @@
     'tower_workflow_launch',
     'tower_job_cancel',
     'tower_workflow_template',
+    'tower_workflow_job_template_schema',
```
This can be removed now right?

john-westcott-iv commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -740,3 +743,179 @@ def wait_on_url(self, url, object_name, object_type, timeout=30, interval=10):
         for k in ('id', 'status', 'elapsed', 'started', 'finished'):
             self.json_output[k] = response['json'].get(k)
+
+    def create_schema_nodes(self, response, schema, workflow_id):
```
Since we only have one module now can we move this logic into the module instead of the tower_api library?

marshmalien committed:
Add execution environment detail to template expanded row

sean-m-sullivan committed:
move methods to module

sean-m-sullivan committed:
move methods to module

sean-m-sullivan committed:
move methods to module

softwarefactory-project-zuul[bot] committed:
Merge pull request #9586 from keithjgrant/8906-code-editor-validation-2

Add VariablesField YAML/JSON validation

SUMMARY
Validates YAML or JSON syntax in the variables field on blur. Prevents formik from submitting the form while this error is present.
addresses #8906, #8907
ISSUE TYPE

Feature Pull Request

COMPONENT NAME

UI

ADDITIONAL INFORMATION

Reviewed-by: Kersom <None>
Reviewed-by: Jake McDermott <yo@jakemcdermott.me>

marshmalien committed:
Fix bug where falsy details return 0

sean-m-sullivan committed:
update lint

softwarefactory-project-zuul[bot] committed:
Merge pull request #9858 from marshmalien/7662-expanded-template-list-item

Add missing template list expanded section details

SUMMARY
Issue: #7662

Add the following detail items:

Description
Organization and link to organization details
Credentials


Move Credential and Label details to the bottom of the expanded section and make them stretch the entire width of the row.


COMPONENT NAME


UI

Reviewed-by: Kersom <None>
Reviewed-by: Tiago Góes <tiago.goes2009@gmail.com>

wenottingham committed:
Update templates for installer changes

softwarefactory-project-zuul[bot] committed:
Merge pull request #9873 from wenottingham/not-any-more-they-arent

Update templates for installer changes

SUMMARY
Some of the options aren't valid any more.

Reviewed-by: Jake McDermott <yo@jakemcdermott.me>

sean-m-sullivan committed:
update lint

sean-m-sullivan committed:
update lint

sean-m-sullivan committed:
remove runit

sean-m-sullivan committed:
remove runit

sean-m-sullivan committed:
linting

softwarefactory-project-zuul[bot] committed:
Merge pull request #9847 from fosterseth/fix_t4922_job_elapsed_time_incorrect

Fix elapsed time on job showing incorrect value

SUMMARY

Elapsed time would always stay at zero

ISSUE TYPE


Bugfix Pull Request

COMPONENT NAME


API

AWX VERSION

awx: 18.0.0

Reviewed-by: Seth Foster <None>
Reviewed-by: Ryan Petrello <None>
Reviewed-by: Alan Rominger <arominge@redhat.com>
Reviewed-by: Jeff Bradberry <None>

beeankha committed:
Change 'ansible-tower-service' to 'automation-controller-service'

softwarefactory-project-zuul[bot] committed:
Merge pull request #9885 from beeankha/remove_old_tower_name_from_docs

Change Reference of 'ansible-tower-service' to 'automation-controller-service' in Docs File

Related to the work done in #9720

Reviewed-by: Shane McDonald <me@shanemcd.com>

ryanpetrello committed:
use flake8 to lint for a few things black doesn't catch

black does *not* warn about missing or extraneous imports,
so let's bring back flake8 in our linting to check for them

softwarefactory-project-zuul[bot] committed:
Merge pull request #9886 from ryanpetrello/autoflake

sprinkle back in some flake8 to catch missing imports

Reviewed-by: Shane McDonald <me@shanemcd.com>
Reviewed-by: Christian Adams <rooftopcellist@gmail.com>
Reviewed-by: Alan Rominger <arominge@redhat.com>
Reviewed-by: Ryan Petrello <None>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9537 from AlexSCorey/lingUIUpdate

Updates Ling UI

SUMMARY
This PR updates Ling ui. One of the reasons for updating this dependency was that they were deprecating withI18n().  They changed their minds on that so we didn't need to urgently remove all those HOCs.  Thus, we can now make that conversion a bit slower a couple of files at a time.
One other thing:  When we are changing the string based on a count of an item (ie. Cancel Job vs. Cancel Jobs) we should use ling ui's <Plural> component. However, in order to show the update strings passed to that component the developer will have to run npm run extract-strings each time they are changed to render the updated strings properly. More info here.
ISSUE TYPE
-dependency  upgrade
COMPONENT NAME

UI

AWX VERSION
ADDITIONAL INFORMATION

Reviewed-by: Kersom <None>
Reviewed-by: Alex Corey <Alex.swansboro@gmail.com>
Reviewed-by: Sergio Moreno <sergiomorenoalbert@gmail.com>

ryanpetrello committed:
record the receptor unit ID on the job record for debugging purposes

see: https://github.com/ansible/awx/issues/8641

softwarefactory-project-zuul[bot] committed:
Merge pull request #7391 from xstasi/devel

Add tower group for collection modules

SUMMARY
It would be useful to be able to declare module defaults for all tower modules, which is accomplished with module groups.
See: https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html
This very simple PR adds the group definition

Reviewed-by: Bianca Henderson <beeankha@gmail.com>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9835 from ryanpetrello/saml-galaxy-cred

automatically setup default Galaxy credentials on SAML login

Reviewed-by: Chris Meyers <None>

softwarefactory-project-zuul[bot] committed:
Merge pull request #9892 from ryanpetrello/record-unit-id

record the receptor unit ID on the job record for debugging purposes

see: #8641

Reviewed-by: Shane McDonald <me@shanemcd.com>

sean-m-sullivan committed:
Merge branch 'tower_workflow_schema' of github.com:sean-m-sullivan/awx into tower_workflow_schema
